tests/federation_room_join_test.go

```diff
@@ -192,3 +195,70 @@ func TestJoinFederatedRoomWithUnverifiableEvents(t *testing.T) {
 		alice.JoinRoom(t, roomAlias, nil)
 	})
 }
+
+// This test checks that users cannot circumvent the auth checks via send_join.
+func TestBannedUserCannotSendJoin(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	srv := federation.NewServer(t, deployment,
+		federation.HandleKeyRequests(),
+		federation.HandleTransactionRequests(nil, nil),
+	)
+	cancel := srv.Listen()
+	defer cancel()
+
+	fedClient := srv.FederationClient(deployment)
+
+	charlie := srv.UserID("charlie")
+
+	// alice creates a room, and bans charlie from it.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	roomID := alice.CreateRoom(t, map[string]interface{}{})
+
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type:     "m.room.member",
+		Sender:   alice.UserID,
+		StateKey: &charlie,
+		Content: map[string]interface{}{
+			"membership": "ban",
+		},
+	})
+
+	// charlie sends a make_join for a different user
+	makeJoinResp, err := fedClient.MakeJoin(context.Background(), "hs1", roomID, srv.UserID("charlie2"), federation.SupportedRoomVersions())
+	must.NotError(t, "MakeJoin", err)
```

Comment:
So Dendrite will fail the test here with `M_FORBIDDEN: eventauth: "@charlie2:host.docker.internal" is not allowed to change their membership from "leave" to "join"` which feels like reasonable behaviour. Annoyingly, Complement CI seems to not be running Dendrite tests through to completion causing this not to be picked up.

Currently Complement tests now fail on the Dendrite repo as a result.

Reply:
I think this should work? it should be a public room so `make_join` should work?

Reply:
The root cause here was that `/createRoom` with `{}` defaults to private rooms as per the spec:

> A public visibility indicates that the room will be shown in the published room list. A private visibility will hide the room from the published room list. Rooms default to private visibility if this key is not included. 

But synapse defaults to public it seems.

Reply:
fixed by #158